extra_data/tests/test_keydata.py

```diff
@@ -169,3 +169,43 @@ def test_drop_empty_trains(mock_sa3_control_data):
     frame_counts = beamview_w_data.data_counts(labelled=False)
     assert frame_counts.shape == (250,)
     assert frame_counts.min() == 1
+
+def test_single_value(mock_sa3_control_data, monkeypatch):
+    f = H5File(mock_sa3_control_data)
+
+    imager = f['SA3_XTD10_IMGFEL/CAM/BEAMVIEW:daqOutput', 'data.image.pixels']
+    flux = f['SA3_XTD10_XGM/XGM/DOOCS', 'pulseEnergy.photonFlux']
+
+    with pytest.raises(NoDataError):
+        imager.as_single_value()  # FEL imager with no data.
+        flux[:0].as_single_value()  # No data through selection.
+
+    # Monkeypatch some actual data into the KeyData object
+    data = np.arange(flux.shape[0])
+    monkeypatch.setattr(flux, 'ndarray', lambda: data)
+
+    with pytest.raises(ValueError):
+        # Try some tolerances that have to fail.
+        flux.as_single_value()
+        flux.as_single_value(atol=1)
+        flux.as_single_value(rtol=0.1)
```

Comment:
N.B. you're only actually testing the first one of these. If you want to test all of them, they each need their own `with pytest.raises` block.

Reply:
(Not precisely the first, of course - it tests that *any* of them raise ValueError, rather than *all* of them.)

Reply:
Oops...